transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,813 @@
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.get_mut()
+            .channel
+            .poll_next_unpin(cx)
+            .map(|e| e.map(Ok))
+    }
+}
+
+impl Transport for Endpoint {
+    type Output = (libp2p_core::PeerId, super::Muxer);
+    type Error = Error;
+    type Listener = Listener;
+    type ListenerUpgrade = Upgrade;
+    type Dial = Upgrade;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
+        let Endpoint(EndpointRef {
+            reference,
+            channel: drop_notifier,
+        }) = self;
+        let socket_addr = if let Ok(sa) = multiaddr_to_socketaddr(&addr) {
+            sa
+        } else {
+            return Err(TransportError::MultiaddrNotSupported(addr));
+        };
+        let own_socket_addr =
+            multiaddr_to_socketaddr(&reference.address).expect("our own Multiaddr is valid; qed");
+        if socket_addr.ip() != own_socket_addr.ip()
+            || (socket_addr.port() != 0 && socket_addr.port() != own_socket_addr.port())
+        {
+            return Err(TransportError::MultiaddrNotSupported(addr));
+        }
+        let channel = reference
+            .receive_connections
+            .lock()
+            .take()
+            .ok_or_else(|| TransportError::Other(Error::AlreadyListening))?;
+        Ok(Listener {
+            reference,
+            drop_notifier,
+            channel,
+        })
+    }
+
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+        let socket_addr = if let Ok(socket_addr) = multiaddr_to_socketaddr(&addr) {
+            if socket_addr.port() == 0 || socket_addr.ip().is_unspecified() {
+                debug!("Instantly refusing dialing {}, as it is invalid", addr);
+                return Err(TransportError::MultiaddrNotSupported(addr));
+            }
+            socket_addr
+        } else {
+            return Err(TransportError::MultiaddrNotSupported(addr));
+        };
+        let Endpoint(endpoint) = self;
+        let mut inner = endpoint.reference.inner.lock();
+        let s: Result<(_, Connection), _> = inner
+            .inner
+            .connect(
+                endpoint.reference.config.client_config.clone(),
+                socket_addr,
+                "l",
+            )
+            .map_err(|e| {
+                warn!("Connection error: {:?}", e);
+                TransportError::Other(Error::CannotConnect(e))
+            });
+        let (handle, connection) = s?;
+        let socket = endpoint.reference.socket.clone();
+        let endpoint = ConnectionEndpoint {
+            pending: socket::Pending::default(),
+            connection,
+            handle,
+            channel: endpoint.channel,
+            waker: None,
+        };
+        Ok(crate::Upgrade::spawn(endpoint, |weak| {
+            inner.muxers.insert(handle, weak);
+            socket
```

Comment:
Why not pass `socket` as a parameter to `spawn`?

Reply:
Fixed